Speedup sparse -> VBR (Fixes #68)

anmolsahoo25 opened the pull request:
Fix #68. On Franz8, runtime down from 60s -> 12s. Also, added a savez parameter to save using numpy (which further saves runtime).

ulysses4ever commented on test.py:
```diff
@@ -9,11 +9,19 @@
 from utils.fileio import write_dense_matrix, write_dense_vector
 from utils.mtx_matrices_gen import vbr_to_mtx
 from utils.utils import extract_mul_nums
+import cProfile
+import timeit
```
Could we add all the new packages to `requirements.txt`? (There are more below.)

anmolsahoo25 replied:
Thanks for the heads-up! I will just remove these, since I don't think we need these in the final codebase.

ulysses4ever commented on utils/convert_real_to_vbr.py:
```diff
@@ -132,14 +139,139 @@ def convert_sparse_to_vbr(mat, rpntr, cpntr, fname, dst_dir):
+
+    return val, indx, bindx, bpntrb, bpntre
+
+def convert_sparse_to_vbr_np(mat, rpntr, cpntr, fname, dst_dir, savez = False):
```
Is it a second version of `convert_sparse_to_vbr`? I'd suggest keeping just one version to avoid all sorts of confusion. This looks like a finer implementation, so let's just switch to it, perhaps?

anmolsahoo25 replied:
I'd lean toward keeping both. It serves as a reference to compare the results of both + fall-back in case the newer version ever has a regression.

ulysses4ever replied:
If we find a regression in the new version, we will be able to fix it. I also don't know why we'd want to compare the results of both: we either can run the comparison now (do you have such experiment ready), before the merge, to make sure we didn't miss any bugs, or we'll find alleged bugs later anyway: a lot depends on this function doing the right thing. So I don't see what purpose the old version serves. Now, the downside of having two versions is clear: the whole code base will be randomly picking which version to use --- this will lead to surprises. E.g. if we assume functional correctness of both, some experiments will run slower than others because we forgot to switch the version. It only creates confusion.

anmolsahoo25 replied:
> I also don't know why we'd want to compare the results of both: we either can run the comparison now (do you have such experiment ready), before the merge,

I am expecting that this function `convert_sparse_to_vbr` will undergo more changes in the future. In that case, it is best to have a version that is easy to check (current `convert_sparse_to_vbr`) to ensure we are not generating the matrix incorrectly. 

> E.g. if we assume functional correctness of both, some experiments will run slower than others because we forgot to switch the version. It only creates confusion.

Don't worry about this. I will edit the code to make sure we are always exporting the correct version under the right alias, so there is no confusion,

reikdas replied:
> I am expecting that this function convert_sparse_to_vbr will undergo more changes in the future. In that case, it is best to have a version that is easy to check (current convert_sparse_to_vbr) to ensure we are not generating the matrix incorrectly.

I don't think this makes sense in a world where we have version control systems. Let's say we have a function `foo`. We make some non-functional changes to `foo`. This does not mean we should now have a `foo_old` and a `foo_new`, so that if `foo_new` fails some day, we can revert to `foo_old`. Instead, we should write strong enough tests for `foo` (in an ideal world, we should have already had these tests before we implemented `foo_new`), and in case we still need to revert to `foo_old` some day, we can use our version control to do that.

ulysses4ever replied:
Well put!

anmolsahoo25 replied:
Agreed, leaving the old version only for the equivalence test.

anmolsahoo25 commented:
As per @ulysses4ever's advice, changed the name of the np version to be standard and updated the reference function to be an internal function. Revert this commit, if it introduces any errors in the rest of the testsuite.

reikdas commented on utils/convert_real_to_vbr.py:
```diff
@@ -132,14 +111,142 @@ def convert_sparse_to_vbr(mat, rpntr, cpntr, fname, dst_dir):
+
+    out_path = os.path.join(dst_dir, f"{fname}.vbr")
+
+    if savez:
```
The problem with saving the VBR in a binary instead of a text file is that the generated C program reads the value array from the text file. Storing it in a binary would imply that we would need `.npz` reading logic in the generated C program.

anmolsahoo25 replied:
Removed savez

reikdas commented on test.py:
```diff
@@ -331,18 +334,50 @@ def test_baselines():
     run_nonzeros_spmv()
     run_nonzeros_spmm()
 
+def test_vbr_conversion():
```
Isn't this test already covered here: https://github.com/reikdas/SABLE/blob/cb10fc6f2576928ec32f1a2b7b9b3b9703d4d357/test.py#L121

anmolsahoo25 replied:
Removed duplicate test.

reikdas commented:
There are a few failing tests: https://github.com/reikdas/SABLE/actions/runs/13346079488/job/37277521173?pr=72 I am concerned that there might be a correctness bug.

anmolsahoo25 commented:
Checked that the failing tasks are passing on my local. Could you run it again? I don't have taskset on my Mac, so its failing some other tests.
